template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmHostGroupRoleConfigProviderProcessor.java

```diff
@@ -45,35 +65,40 @@ public void process(CmTemplateProcessor templateProcessor, TemplatePreparationOb
         List<ApiClusterTemplateHostTemplate> hostTemplates = getHostTemplates(templateProcessor);
         Map<String, ServiceComponent> serviceComponents = templateProcessor.mapRoleRefsToServiceComponents();
 
         for (ApiClusterTemplateHostTemplate hostTemplate : hostTemplates) {
             String hostGroupName = hostTemplate.getRefName();
             List<String> roleConfigGroups = ofNullable(hostTemplate.getRoleConfigGroupsRefNames()).orElseGet(List::of);
             HostgroupView hostgroupView = hostGroups.get(hostGroupName);
+            groupByHostGroupName(source, configsByRoleConfigGroup, serviceComponents, hostGroupName, roleConfigGroups, hostgroupView);
+        }
+        return configsByRoleConfigGroup;
+    }
 
-            for (String roleConfigGroup : roleConfigGroups) {
-                for (CmHostGroupRoleConfigProvider provider : providers) {
-                    ServiceComponent serviceComponent = serviceComponents.get(roleConfigGroup);
-
-                    if (serviceComponent != null
-                            && Objects.equals(provider.getServiceType(), serviceComponent.getService())
-                            && provider.getRoleTypes().contains(serviceComponent.getComponent())) {
-
-                        configsByRoleConfigGroup.computeIfAbsent(roleConfigGroup, __ -> new HashMap<>())
-                                .computeIfAbsent(hostGroupName, __ -> new ArrayList<>())
-                                .addAll(provider.getRoleConfigs(serviceComponent.getComponent(), hostgroupView, source));
-                    }
+    private void groupByHostGroupName(TemplatePreparationObject source, Map<String, Map<String, List<ApiClusterTemplateConfig>>> configsByRoleConfigGroup,
+            Map<String, ServiceComponent> serviceComponents, String hostGroupName, List<String> roleConfigGroups, HostgroupView hostgroupView) {
+        for (String roleConfigGroup : roleConfigGroups) {
+            for (CmHostGroupRoleConfigProvider provider : providers) {
+                ServiceComponent serviceComponent = serviceComponents.get(roleConfigGroup);
+                if (serviceComponent != null
+                        && isSameByServiceTypeAndRoleType(provider, serviceComponent)
+                        && !serviceComponent.getComponent().equals(sharedComponents.get(serviceComponent.getService()))) {
+                    Map<String, List<ApiClusterTemplateConfig>> configs = configsByRoleConfigGroup.computeIfAbsent(roleConfigGroup, __ -> new HashMap<>());
+                    configs.computeIfAbsent(hostGroupName, __ -> new ArrayList<>())
+                            .addAll(provider.getRoleConfigs(serviceComponent.getComponent(), hostgroupView, source));
                 }
             }
         }
-        return configsByRoleConfigGroup;
+    }
+
+    private boolean isSameByServiceTypeAndRoleType(CmHostGroupRoleConfigProvider provider, ServiceComponent serviceComponent) {
```

Comment:
maybe isEqual ?

Reply:
changed

Reply:
is this better? `isEqualsByServiceTypeAndRoleType`

Reply:
I think so yes

Reply:
ok, changed